[Inductor CUTLASS backend] Step 5: CUTLASS gemm kernels

ipiszy opened the pull request:
This is the step 5 to add cutlass as an alternative inductor backend.

Feature request: https://github.com/pytorch/pytorch/issues/106991.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #107933
* #107931
* #107901
* #107847
* #107802



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

ipiszy committed:
[Inductor CUTLASS backend] Step 5: CUTLASS gemm kernels

[ghstack-poisoned]

kadeng commented on test/inductor/test_max_autotune.py:
```diff
@@ -188,6 +188,64 @@ def mm(a, b):
+        ):
+            Y_compiled = torch.compile(mm, dynamic=dynamic)(a, b)
+            Y = mm(a, b)
+            torch.testing.assert_close(Y_compiled, Y)
```
I think we should somehow make sure here that the function under test is actually being called and not, for example, a Triton variant.

aakhundov replied:
Agreed. `run_and_get_code` may be helpful ([example](https://github.com/pytorch/pytorch/blob/d4a99631dd589afbb972b3401e56f029192c9b0f/test/inductor/test_pattern_matcher.py#L62) from other test).

ipiszy replied:
Makes sense, updated max_autotune_gemm_backends to ensure that there is a test case for "CUTLASS" only.

kadeng commented on torch/_inductor/codegen/cuda/gemm_template.py:
```diff
@@ -0,0 +1,472 @@
+extern "C" {
+{{kernel.def_kernel(inputs=[X, W, Bias], outputs=[Y], names_str="X, W, Bias, Y", input_reorder=input_reorder)}} {
+  try {
+  // printf("X: %p\n", (void*)(X));
```
We could conditionally generate these in if (in an uncommented manner) if some debug flag is set, otherwise I think they should not become part of the generated code.

ipiszy replied:
Yes didn't fix it in draft code ^^.

kadeng commented on torch/_inductor/codegen/cuda/gemm_template.py:
```diff
@@ -0,0 +1,472 @@
+                num_3x_ops += 1
+            else:
+                num_2x_ops += 1
+        print(f"Got cutlass configs: {len(res)=}, {num_3x_ops=}, {num_2x_ops=}")
```
Leftover print statement from development?

ipiszy replied:
Yes didn't fix it in draft code ^^.

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 5: CUTLASS gemm kernels"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 5: CUTLASS gemm kernels"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 5: CUTLASS gemm kernels"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 5: CUTLASS gemm kernels"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 5: CUTLASS gemm kernels"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 5: CUTLASS gemm kernels"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 5: CUTLASS gemm kernels"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 5: CUTLASS gemm kernels"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 5: CUTLASS gemm kernels"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 5: CUTLASS gemm kernels"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 5: CUTLASS gemm kernels"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy reviewed (commented):
Thanks @kadeng !

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 5: CUTLASS gemm kernels"

This is the step 5 to add cutlass as an alternative inductor backend.

Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "[Inductor CUTLASS backend] Step 5: CUTLASS gemm kernels"

This is the step 5 to add cutlass as an alternative inductor backend.

Feature request: https://github.com/pytorch/pytorch/issues/106991.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy commented:
@kadeng @aakhundov I'm sorry I messed up the PR stack and somehow ghstack recreated a new PR for this: https://github.com/pytorch/pytorch/pull/108015. Let's move to this PR for discussions.
